### ISSUE TITLE
Geofence notification - add script call in addition to mail/web

### ISSUE THREAD

**Issue by TwizzyDizzy** [tananaev/traccar#2101]
Hi Anton,

first and foremost: thanks for the great new feature "geofencing". This is highly appreciated!

What I was thinking of: in order to keep the notification stuff generic, one could add the option to call an external resource. This could be:

* a URI which receives the relevant information via POST request
* a script call on the local server (this would be the most generic version and would enable nearly any kind of integration)

The crucial part would be to transmit the data in a generic fashion so the external script could utilize them.

Cheers
Thomas

**Comment by tananaev** [comment]
It would be possible to have something similar to data forwarding that we have, but for now I guess email is pretty generic way.

**Comment by TwizzyDizzy** [comment]
Unfortunately with mail it's not really possible to trigger something external with just this hook :|

**Comment by tananaev** [comment]
@Abyss777, what do you think? How hard would it be to add another option to forward event to a URL (something similar to `forward.enable` and `forward.url`).

**Comment by Abyss777** [comment]
It is in my nearest plans. Only question, what we will forward? event object (as json for example) or preformated message?
Forwarding to script is insecure, we have to sonsider it.

**Comment by TwizzyDizzy** [comment]
I'd suggest sending a serialized object instead of a preformatted message. The former is much more easy to handle. Preformatted messages would lead to parsing this message which is simply not neccessary.

Cheers
Thomas

**Comment by tananaev** [comment]
Parsing form parameters should be even easier than JSON.

**Comment by Abyss777** [comment]
We can even POST only event id. Details can be retrieved via API.
Actually other side will request `position` after receiving event in most cases to get whole information.
Or we can POST two objects: event and position, as we do for web-client.

1. Just event id
2. Only event via parameters (not json)
3. Event and position (if need)

What is better?

**Comment by tananaev** [comment]
I guess JSON would be easier to implement, so we can go with that option.

**Comment by TwizzyDizzy** [comment]
My two cents: From a scaling perspective (assuming a huge install of Traccar) it would probably be better to send as much information as possible with the request so the server doesn't have to handle additional API requests because some information in the notification was missing.

Cheers
Thomas

**Comment by tananaev** [comment]
I agree that it would be better to send all information straight away.

**Comment by Abyss777** [comment]
OK, I see your points.

**Comment by TwizzyDizzy** [comment]
Just for my understanding. How would one configure this? From my understanding, this should be configurable on a per user basis, because user A might want his notifications to go to ServerY and user B might want his notifications to go to ServerZ.

Am I missing something?

**Comment by Abyss777** [comment]
No, It is unconditional forwarding all events to other server (service). It can't be configured per user.
It is helpful for situation when traccar is a part of some bigger system.

Can you give an example when different users of one traccar server want to send RAW objects to other servers? They have to implement his own web-server, parsers and other things.

It will be useful to send preformated messages configured per user, for example to some push service. I thought about it, but It have to be very configurable because of zoo of services.
For example some push services use GET, some POST, some get auth token via request, some via header and so on.
It hard to do without template framework. It is in plans.

**Comment by TwizzyDizzy** [comment]
Ah... I remember traccar having a fleet tracking background, right? From that perspective there is, indeed, little sense to implement it that way.

My background, on the other hand, is that I self-host most of my software and sometimes give accounts in my instance to others. In that scenario, others might want to push their messages to another server than mine whereas I want to push the messages to my server. Just think of how it would have to be, if traccar was a SaaS somebody is offering to others rather than an internal application for a company.

And you are right. The other side, of course, has to implement / adapt to the message traccar is pushing. I **don't** think that traccar should offer "presets" or so to push to Service X oder Service Y. It should only commit to "traccar sends geofence notifications in this or that way. deal with it, adapt to it.".

Cheers
Thomas

**Comment by tananaev** [comment]
Thanks for feedback. We can definitely consider per-user option in future, but I think for now there are some other higher priority items in the backlog.

**Comment by TwizzyDizzy** [comment]
Yeah... that's fine, I just meant it as a "nice to have"-feature for the future :)
